File: image-analyzer/image-parser.py
import os
import json
import argparse
from pathlib import Path
from openai import OpenAI
from typing import Dict, List
import base64
import datetime
import time
import sys

def check_api_key():
    """Check if OpenAI API key is set and valid."""
    api_key = os.getenv('OPENAI_API_KEY')
    if not api_key:
        raise ValueError(
            "OpenAI API key not found! Please set your OPENAI_API_KEY environment variable.\n"
            "You can do this by running:\n"
            "   set OPENAI_API_KEY=your-api-key-here    (on Windows)\n"
            "   export OPENAI_API_KEY=your-api-key-here  (on Unix/Linux)"
        )
    return api_key

def analyze_image(client: OpenAI, image_path: str) -> Dict:
    """Analyze a single image using OpenAI's Vision model."""
    try:
        with open(image_path, "rb") as image_file:
            # Convert image to base64
            base64_image = base64.b64encode(image_file.read()).decode('utf-8')
            
            response = client.responses.create(
                model="gpt-4o-mini",
                input=[{
                    "role": "user",
                    "content": [
                        {"type": "input_text", "text": "Please extract all mathematical and MATLAB content shown in this image. Include formulas, code snippets, and technical details exactly as they appear. For example images in the images, extract them as alt text."},
                        {
                            "type": "input_image",
                            "image_url": f"data:image/png;base64,{base64_image}"
                        }
                    ]
                }]
            )
            
            return {
                "filename": os.path.basename(image_path),
                "description": response.output_text,
                "timestamp": str(datetime.datetime.now())
            }
    except Exception as e:
        print(f"Error processing {image_path}: {str(e)}")
        return {
            "filename": os.path.basename(image_path),
            "error": str(e),
            "timestamp": str(datetime.datetime.now())
        }

def process_directory(input_dir: str, output_dir: str) -> None:
    """Process all images in the input directory and save results to output directory."""
    input_path = Path(input_dir)
    output_path = Path(output_dir)
    
    # Validate input directory
    if not input_path.exists():
        raise ValueError(f"Input directory '{input_dir}' does not exist")
    if not input_path.is_dir():
        raise ValueError(f"'{input_dir}' is not a directory")
        
    # Create output directory if it doesn't exist
    output_path.mkdir(parents=True, exist_ok=True)
    
    # Check API key first
    api_key = check_api_key()
    
    # Initialize OpenAI client with the API key
    client = OpenAI(api_key=api_key)
    
    # Process each image
    image_files = sorted(input_path.glob("*.png"))
    if not image_files:
        print(f"No PNG images found in {input_dir}")
        return
        
    print(f"Found {len(image_files)} PNG images to process")
    for image_file in image_files:
        print(f"Processing {image_file.name}...")
        
        # Analyze image
        result = analyze_image(client, str(image_file))
        
        # Save result to JSON
        output_file = output_path / f"{image_file.stem}.json"
        with open(output_file, 'w', encoding='utf-8') as f:
            json.dump(result, f, indent=2, ensure_ascii=False)
        
        print(f"Saved analysis to {output_file}")
        # Add a small delay to avoid rate limiting
        time.sleep(1)

def main():
    parser = argparse.ArgumentParser(description='Analyze images using OpenAI Vision API')
    parser.add_argument('--input-dir', type=str, default="images/Chapter 2 - Basic Use of MatLab",
                      help='Directory containing images to analyze')
    parser.add_argument('--output-dir', type=str, default="chapter-1/data",
                      help='Directory to save JSON outputs')
    
    args = parser.parse_args()
    try:
        process_directory(args.input_dir, args.output_dir)
    except Exception as e:
        print(f"Error: {str(e)}")
        return 1
    return 0

if __name__ == "__main__":
    sys.exit(main()) 

File: image-analyzer/image-parser.js
import OpenAI from 'openai';
import { readFileSync, writeFileSync, readdirSync, mkdirSync, existsSync } from 'fs';
import { join, basename } from 'path';
import dotenv from 'dotenv';
import { fileURLToPath } from 'url';
import { dirname } from 'path';

// Load environment variables from .env
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
dotenv.config({ path: join(__dirname, '../.env') });

// Initialize OpenAI client with API key from .env
const openai = new OpenAI({
    apiKey: process.env.OPENAI_API_KEY
});

// Analyze a single image
const analyzeImage = async (imagePath) => {
    try {
        // Read and convert image to base64
        const imageBuffer = readFileSync(imagePath);
        const base64Image = imageBuffer.toString('base64');

        const response = await openai.responses.create({
            model: "gpt-4o-mini",
            input: [{
                role: "user",
                content: [
                    { 
                        type: "input_text", 
                        text: `You are a text extractor from images. Please extract all the content shown in this image and output it in this specific json format. For example images in the images, extract them as alt text.

{
    "section": {
        "title": "string",  // The section title (e.g., "2.1 Introduction")
        "content": {
            "main_concept": {
                "term": "string",      // The main term being defined (e.g., "MATLAB")
                "definition": "string" // The complete definition
            },
            "key_points": [
                {
                    "topic": "string",     // Topic name (e.g., "toolboxes", "command", etc.)
                    "description": "string" // Detailed explanation
                }
            ],
            "examples": {
                "commands": [
                    {
                        "code": "string",    // The actual command
                        "explanation": "string" // Optional explanation if provided
                    }
                ],
                "images": [
                    {
                        "alt_text": "string",    // Description of any example images shown
                        "context": "string"      // Where/how the image is used in the content
                    }
                ]
            },
            "data_types": [
                {
                    "name": "string",      // Name of the data type
                    "description": "string" // Description/explanation
                }
            ]
        }
    }
}

Extract all text content and format it according to this structure. Preserve exact code snippets and technical terms as they appear in the image. If there are any example images shown in the content, describe them in the examples.images array.`
                    },
                    {
                        type: "input_image",
                        image_url: `data:image/png;base64,${base64Image}`
                    }
                ]
            }]
        });

        return {
            filename: basename(imagePath),
            description: response.output_text,
            timestamp: new Date().toISOString()
        };
    } catch (error) {
        console.error(`Error processing ${imagePath}:`, error.message);
        return {
            filename: basename(imagePath),
            error: error.message,
            timestamp: new Date().toISOString()
        };
    }
};

// Process all images in a directory
const processDirectory = async (inputDir, outputDir) => {
    // Validate input directory
    if (!existsSync(inputDir)) {
        throw new Error(`Input directory '${inputDir}' does not exist`);
    }

    // Create output directory if it doesn't exist
    if (!existsSync(outputDir)) {
        mkdirSync(outputDir, { recursive: true });
    }

    // Get all PNG files
    const imageFiles = readdirSync(inputDir)
        .filter(file => file.toLowerCase().endsWith('.png'))
        .sort();

    if (imageFiles.length === 0) {
        console.log(`No PNG images found in ${inputDir}`);
        return;
    }

    console.log(`Found ${imageFiles.length} PNG images to process`);

    // Process each image
    for (const imageFile of imageFiles) {
        console.log(`Processing ${imageFile}...`);
        
        const imagePath = join(inputDir, imageFile);
        const result = await analyzeImage(imagePath);

        // Save result to JSON
        const outputFile = join(outputDir, `${basename(imageFile, '.png')}.json`);
        writeFileSync(outputFile, JSON.stringify(result, null, 2), 'utf8');
        
        console.log(`Saved analysis to ${outputFile}`);
        
        // Add a small delay to avoid rate limiting
        await new Promise(resolve => setTimeout(resolve, 1000));
    }
};

// Main function
const main = async () => {
    const inputDir = process.argv[2] || "../images/Chapter 1 - Intro_editedversion_freeElective_report";
    const outputDir = process.argv[3] || "../chapter-1/data";

    try {
        await processDirectory(inputDir, outputDir);
    } catch (error) {
        console.error("Error:", error.message);
        process.exit(1);
    }
};

// Run the script
main(); 